### ISSUE TITLE
LB IPAM Service Affinity

### ISSUE THREAD

**Issue by CompPhy** [cilium/cilium#33440]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

We are deployed into a small Talos cluster running on VMWare.  This is using LB IPAM to announce ingress services to our network.  We do have other services working fine here, but this is the first one that's needing serviceAffinity which is where the issue appears to lie.

The symptoms are that we have a DaemonSet which is doing HTTP Auth Digest.  This requires session affinity to direct the initial connection to the same pod, but that does not appear to be happening.  Normally with this type of authentication you see two initial connections, the first is a 401 with a nonce and then the second contains the real authentication information.  In our case we are seeing these two connections go to different pods, which results in a failed authentication.

We have `enable-session-affinity` set to `true`, and the ingress/service are responding.  However, I noticed on the service that session affinity is not being set.  Since the services are automatically generated through Cilium, I'm not really sure how to affect this.  The assumption that's what the `enable-session-affinity` option is for.

We were also able to verify that the problem goes away when we set the application scaling to only have one replica.  If there's only a single replica and the traffic always goes to the same pod, then everything works as expected and we get proper authentication.

### Cilium Version

cilium-cli: v0.16.4 compiled with go1.22.1 on linux/amd64
cilium image (default): v1.15.3
cilium image (stable): v1.15.6
cilium image (running): 1.15.6

### Kernel Version

Linux version 6.1.80-talos (@buildkitsandbox) (gcc (GCC) 13.2.0, GNU ld (GNU Binutils) 2.41)

### Kubernetes Version

Client Version: v1.29.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.29.2

### Sysdump

My sysdump file is larger than the 25MB limit on GitHub.  If this is really needed, I will need to know how to reduce it's size.

### Relevant log output

```shell
I will include a screenshot of the application logs below.  You can clearly see that it sent to two consecutive requests to two different pods.

Here is the service config created by LB IPAM, you can see that service affinity is set to None.  This is after verifying that `enable-session-affinity` is set to true.

Name:                     cilium-ingress-datareceiver
Namespace:                datareceiver
Labels:                   cilium.io/ingress=true
Annotations:              <none>
Selector:                 <none>
Type:                     LoadBalancer
IP Family Policy:         SingleStack
IP Families:              IPv4
IP:                       10.109.173.255
IPs:                      10.109.173.255
LoadBalancer Ingress:     10.10.90.13
Port:                     http  80/TCP
TargetPort:               80/TCP
NodePort:                 http  30540/TCP
Endpoints:                
Port:                     https  443/TCP
TargetPort:               443/TCP
NodePort:                 https  32136/TCP
Endpoints:                
Session Affinity:         None
External Traffic Policy:  Cluster
Events:                   <none>
```

### Anything else?

![Screenshot 2024-06-27 154045](https://github.com/cilium/cilium/assets/10223024/c5278a5d-9508-477f-97a4-645041d8d766)

### Cilium Users Document

- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by youngnick** [comment]
Yes, the `enable-session-affinity` setting tells Cilium to support the `sessionAffinity` directive in Service objects.

However, for Ingress session affinity support, there are actually two places that need to support session affinity: the Loadbalancer Service that's created by Cilium, and Cilium's Envoy config.

We can change Cilium to add `sessionAffinity` to the Loadbalancer Services it creates if `enable-session-affinity` is enabled, but that will just make the traffic for a particular client go to the same Envoy. Envoy also needs config to enable session affinity, which we currently don't have a way to configure. Most times I've spoken to people about this, they want to be able to do it on a per-Ingress basis, which means that we'll need a way to configure this per Ingress.

Gateway API does have an experimental way to configure this, so I'd prefer to put engineering effort towards supporting that method rather than adding _more_ annotations to Cilium's Ingress support (which is the only viable method to add per-Ingress config due to the spec's shortcomings).

In summary, you're right, Ingress does not currently support session affinity, but adding it is not as straightforward as just adding `sessionAffinity` to the Loadbalancer Services. Without an ability to add session affinity config to Envoy as well, making that change won't do anything.

I can't see a CFP issue for adding session affinity however, so we can leave this as a request for the feature if you like.

**Comment by CompPhy** [comment]
Thank you for the explanation.  Yes, please do consider this a feature request.

Let me make sure I'm following... We can set serviceAffinity on service objects and it should work as expected.  However, when using cilium ingress, which creates it's own load balance service, that doesn't support creating the annotation on it's own service object.

I think I get the explanation as to why here, but it seems like the current feature set is incomplete.  Session affinity isn't a new concept and many web services rely on it for consistent end user experience.  I can think of many ways that service layer is not sufficient here, the most obvious example being that ingress provides for TLS or advanced URL routing; which cannot be provided at the service layer.

I get that the long term goal is Gateway API, but Ingress has been around for a long time and the switch isn't trivial.  Looking at many common helm applications, most don't even support Gateway API yet because it's still new.  Didn't it just hit 1.0 in very recent months?

This posses a question as well.... Is there a way for us to create our own LoadBalance service, which does support the annotation, and then cilium ingress uses that instead of creating it's own separate service???  If a load balanced service is already provided, maybe this is a shorter path to get us what we need?  As it is right now, if we manually create the service, marked with LoadBalance, cilium still creates a second service as part of ingress.  I've looked through documentation and it doesn't seem like it's possible to bypass this second service layer, but maybe I missed something?

**Comment by youngnick** [comment]
Okay, I think that some parts of how Cilium's Ingress support works are not clear, so let me try to explain.

Cilium's Ingress support works by intercepting traffic bound for a Loadbalancer Service that is associated with Cilium and forwarding that traffic to the local Envoy instance. Once the traffic arrives at Envoy, Envoy has its own loadbalancing setup that chooses a Service backend to send the traffic to. So, for any request coming in to an Ingress, there are _two_ load balancing decision points in the traffic flow.

Setting serviceAffinity on the Loadbalancer service will ensure that the load balancer that is outside the cluster will always route traffic for a matching flow (usually, this is done on the 5-tuple of source address, source port, destination address, destination port, and protocol) to the same endpoint ( which in Cilium Ingress's case is one of the Kubernetes Nodes). We can definitely make this change, and have Cilium's Ingress controller use serviceAffinity. But that will only solve half of the problem.

We still need to be able to configure the Envoy load balancing so that, when the traffic arrives at the same Envoy (because of the serviceAffinity on the created Loadbalancer Service), the traffic is then directed to the same backend _by Envoy_. Currently, we don't configure any session affinity inside Envoy, so I can't be confident that this would "just work" without trying it out.

Hopefully that makes this more clear.

In summary:
- we can add support for serviceAffinity to the created Loadbalancer Services
- it may not actually give you full service affinity to the backends (which I assume is what you want)

Aside from that, there is the question of how we configure the Ingress controller to set serviceAffinity. I think the first place would be to set serviceAffinity on the shared Loadbalancer Service that is created by the Helm chart if serviceAffinity is enabled, which would mean that all Ingress objects that use the shared Ingress LB Service would be affected.

Do you use shared or Dedicated Ingress, @CompPhy?

**Comment by CompPhy** [comment]
Yeah, I guess the Envoy part slipped my mind.  I can see how this might need more work.  Although, if you take the approach of Nginx or AWS LB here, which use cookies to track sessions, maybe it becomes a little easier?  Search for sticky sessions if you don't know what I'm referring to here.

We're using dedicated, although really one would expect that either mode should work.  I believe we tested shared at one point but ended up switching because we want to have dedicated IP's for several of the services.

I do have a perhaps related related question here.   I noticed that if/when there is an already existing LoadBalancer type service, cilium still creates it's own service for the ingress.  It seems kind of redundant to create a second service in this case if there's an already existing service that's of type LoadBalancer.  In the case of LB IPAM, this also results in two IP addresses being assigned; one for the original service and one for the cilium service.  In cases like this, shouldn't it be possible for cilium to map the existing service instead of making another one?  Maybe this isn't possible because of how Envoy interacts here?

**Comment by youngnick** [comment]
I know this may sound counterintuitive, but basically no Ingress controller will do sticky sessions or session affinity without some config. ingress-nginx uses annotations, and in Gateway API, we're also building out structured field support for configuring how to choose the affinity you require. (Turns out there is no clear standard here, so everyone does something slightly different). To be honest, it's unlikely that we would build this for Ingress, but we will implement the Gateway API version of the config once it's available.

To answer your related question somewhat indirectly:

Cilium has two load balancer modes for Ingress: shared and dedicated. For Shared mode, the Helm chart provisions a Loadbalancer Service at install time (named `cilium-ingress`, in the install namespace, which is usually `kube-system`), and Ingress config will be exposed on that Service. For Dedicated mode, each Ingress will generate a separate load balancer, and expose the Ingress on that Service's LB. We offer Shared mode precisely to mitigate the problem you're talking about.

We have to have a Cilium specific Service in both cases because we need to be able to forward traffic that arrives at nodes bound for that Service to Envoy using eBPF - we can't do that on just any Loadbalancer Service.

**Comment by jordangarside-kiwi** [comment]
Has there been any progress on sticky sessions for the Gateway API?

I like the Gateway API, but I'm thinking I'm going to have to install the nginx ingress controller and stop using the cilium-envoy ingress controller because the Gateway API still can't quite replace ingresses yet.

**Comment by youngnick** [comment]
There is a BackendLBPolicy object upstream which allows settings that can configure sticky sessions. It's still Experimental, and Cilium hasn't got around to building out the support yet, but we do intend to.

**Comment by borutlukic** [comment]
Is there any timeline on this? 

We are having the same issue. Any Atlassian product requires cookie session affinity. 
Our current solution is to run a single nginx pod for ingress, which works, but is not really HA setup. 

I was thinking maybe something could be done via CiliumClusterwideEnvoyConfig/CiliumEnvoyConfig  CRD? But I can not find any documentation for that. And then I noticed that in extensions_build_config.bzl the required extension is disabled anyway.

**Comment by youngnick** [comment]
Sorry no, no current timeline. As always, PRs welcomed, although this is big enough that if anyone wants to pick it up, I'd recommend checking in first to get some tips about where to start.

**Comment by vgallegob** [comment]
I also have this issue. Blazor applications require sticky sessions for websocket reconnection.

Perhaps explainig this in the docs would spare us the trouble of digging why we can define it but it doesen't have any effect.

What would you recomend us do for those using Cilium that need sticky sessions with cookies?

**Comment by youngnick** [comment]
As of today, we don't have any way to do sticky sessions with cookies in Cilium's Ingress or Gateway API support. There is an experimental BackendTrafficPolicy object that we could implement that will help, but no contributor has stepped up to actually implement it yet. As always, PRs welcomed, but until that time, Cilium cannot meet sticky sessions with cookies needs.

**Comment by youngnick** [comment]
I should note that an initial version of Session Persistence has been implemented in #48029. See #47089 for more info.
